### PR TITLE
Fix: check destructuring for "no-shadow-restricted-names" (fixes #10467)

### LINKED ISSUE #10467: no-shadow-restricted-names doesn't check destructuring
**Tell us about your environment**

* **ESLint Version:** master
* **Node Version:** -
* **npm Version:** -

**What parser (default, Babel-ESLint, etc.) are you using?**

- default

**Please show your full configuration:**
**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
/*eslint no-shadow-restricted-names: error */

function f1() {
    var undefined = 1
}

function f2() {
    var [undefined] = [1]
}

function f3() {
    var {undefined} = {undefined: 1}
}
```

[Online demo](https://eslint.org/demo/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXM6IGVycm9yICovXHJcblxyXG5mdW5jdGlvbiBmMSgpIHtcclxuICAgIHZhciB1bmRlZmluZWQgPSAxXHJcbn1cclxuXHJcbmZ1bmN0aW9uIGYyKCkge1xyXG4gICAgdmFyIFt1bmRlZmluZWRdID0gWzFdXHJcbn1cclxuXHJcbmZ1bmN0aW9uIGYzKCkge1xyXG4gICAgdmFyIHt1bmRlZmluZWR9ID0ge3VuZGVmaW5lZDogMX1cclxufVxyXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjgsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**What did you expect to happen?**

There are three warnings.

**What actually happened? Please include the actual, raw output from ESLint.**

There is a warning.

----

- I found it in making #10429.
- Maybe we can merge that rule to the [no-shadow](https://eslint.org/docs/rules/no-shadow) rule.

### PR DESCRIPTION
**Pull request by g-plane**
**What is the purpose of this pull request? (put an "X" next to item)**

[x] Bug fix

**What changes did you make? (Give an overview)**
fix #10467 

**Is there anything you'd like reviewers to focus on?**
@mysticatea said that this rule can be merged into the `no-shadow` rule, but I think this should be done in another commit/PR.

### CODE REVIEW

File: tests/lib/rules/no-shadow-restricted-names.js [hunk 1]
```diff
@@ -89,6 +89,21 @@ ruleTester.run("no-shadow-restricted-names", rule, {
                 { message: "Shadowing of global property 'eval'.", type: "Identifier" },
                 { message: "Shadowing of global property 'eval'.", type: "Identifier" }
             ]
+        },
```
Comment (Aladdin-ADD) [inline tests/lib/rules/no-shadow-restricted-names.js]: could you please add some more tests like:
```js
var undefined = "foo";
var foo = bar, undefined = bar;
```

Comment (g-plane) [reply tests/lib/rules/no-shadow-restricted-names.js]: What's the different between the existing tests?

Comment (Aladdin-ADD) [reply tests/lib/rules/no-shadow-restricted-names.js]: oh, my apologies! I didn't notice it was already there!

**Comment by mysticatea** [comment]
Thank you for the contributing!

However, I'd like to point out some matters.

- It should handle destructuring in `checkForViolation` function rather than `VariableDeclarator` handler because function parameters can have it as well.
- It should handle the following things as well:

  - nested destructuring (e.g. `var {a: {b: {undefined}}} = obj`).
  - renaming in destructuring (e.g. `var {a: undefined} = obj`).
  - default values (e.g. `var {a: undefined = 0} = obj`).
  - rest elements (e.g. `var {a, ...undefined} = obj`).

  So I'd like to recommend to use `context.getDeclaredVariables` API.

- `Array.prototype.map` is the method which converts an array to another array. In this case `Array.prototype.forEach` is more proper.
- #10429 modifies the same file, so I'm worried to need resolve confliction later.

**Comment by g-plane** [comment]
Can I replace `[].call.map` in the rule file?

**Comment by g-plane** [comment]
Updated.

**Comment by mysticatea** [comment]
Looks good. But it can be simplified more:

```js

        return {
            "VariableDeclaration, :function, CatchClause"(node) {
                for (const variable of context.getDeclaredVariables(node)) {
                    if (variable.defs.length > 0 && RESTRICTED.includes(variable.name)) {
                        context.report({
                            node: variable.defs[0].name,
                            message: "Shadowing of global property '{{idName}}'.",
                            data: {
                                idName: variable.name
                            }
                        });
                    }
                }
            }
        };
```

**Comment by g-plane** [comment]
Updated again.

**Comment by platinumazure** [comment]
Merged, thanks @g-plane for contributing!

**Pull request merged**